task-maker-format/src/ioi/sanity_checks/task.rs

```diff
@@ -121,3 +125,95 @@ impl SanityCheck for NoBitsStdCpp {
         Ok(())
     }
 }
+
+/// Check that "implied" subtasks dependecies do not form cycles.
+#[derive(Debug, Default)]
+pub struct SubtaskDependencies;
+make_sanity_check!(SubtaskDependencies);
+
+impl SanityCheck for SubtaskDependencies {
+    type Task = IOITask;
+
+    fn name(&self) -> &'static str {
+        "SubtaskDependencies"
+    }
+
+    fn category(&self) -> SanityCheckCategory {
+        SanityCheckCategory::Task
+    }
+
+    fn post_hook(&self, task: &Self::Task, eval: &mut EvaluationData) -> Result<(), Error> {
+        let mut table = HashSet::new();
+        for &st1 in task.subtasks.keys() {
+            for &st2 in task.subtasks.keys() {
+                if st1 != st2 {
+                    table.insert((st1, st2));
+                }
+            }
+        }
+
+        for sol in &eval.solutions {
+            let mut map = HashMap::new();
+            for check in &sol.checks {
+                let val = match check.result {
+                    SolutionCheckResult::Accepted => 1.0,
+                    SolutionCheckResult::PartialScore => f32::NAN,
+                    _ => 0.0,
+                };
+
+                for subtask in task.find_subtasks_by_pattern_name(&check.subtask_name_pattern) {
+                    map.insert(subtask.id, val);
+                }
+            }
+
+            for (&k1, v1) in &map {
+                for (&k2, v2) in &map {
+                    if matches!(v1.partial_cmp(v2), None | Some(Ordering::Greater)) {
+                        table.remove(&(k1, k2));
+                    }
+                }
+            }
+        }
+
+        'outer: for &st1 in task.subtasks.keys() {
+            let mut sts = vec![st1];
+            for &st2 in task.subtasks.keys() {
+                if table.contains(&(st1, st2)) && table.contains(&(st2, st1)) {
+                    if st1 < st2 {
+                        sts.push(st2);
+                    } else {
+                        continue 'outer;
+                    }
+                }
+            }
+            if sts.len() > 1 {
+                sts.sort();
+                eval.add_diagnostic(
+                    Diagnostic::warning(format!(
+                        "Subtasks {sts:?} are solved by the same set of solutions",
```

Comment:
It would be nice to use the subtask names if they are available, not just the indexes.